### ISSUE TITLE
✨ Allow $schema field in dagger.json

### ISSUE THREAD

**Issue by JacobLey** [dagger/dagger#8967]
### What are you trying to do?

I am trying to add a `$schema` field to my `dagger.json` file.

e.g.
```json
{
  "$schema": "https://docs.dagger.io/reference/dagger.schema.json",
  "name": "my-module",
  "engineVersion": "v0.14.0",
  "sdk": "go",
  "source": "."
}
```

Currently my IDE flags the `$schema` field as not allowed, because it is not explicitly declared [in the schema](https://docs.dagger.io/reference/dagger.schema.json).

See how other JSON schemas support this, like [JSON schema's meta schema itself](https://json-schema.org/draft/2020-12/meta/core).

The field is also deleted when I run `dagger develop`. I would expect it to be retained, to the exact same URL as before (i.e. perhaps I want a specific version's schema)

### Why is this important to you?

Whenever I am working with a new piece of software/framework, there is a fantastic amount of discoverability/self-documentation available in config files when an IDE can auto-populate descriptions based on a `$schema` field.

I don't necessarily expect Dagger itself to enforce the schema abides by the specified `$schema` URL (can continue validating however it currently does based on current dagger version, with exception of allowing `$schema`).
I am primarily using the field for my own developer experience.

### How are you currently working around this?

Manually adding in the `$schema` field when I want to make edits, to ensure I am using the right field names/updating the right settings.

**Comment by JacobLey** [comment]
Naively, editing the JSON Schema file to support this field is trivial. Would be happy to open a PR for that if agreed upon.

The trickier part is the "don't delete my `$schema` if it exists" part

**Comment by JacobLey** [comment]
Took a swing at a PR: https://github.com/dagger/dagger/pull/9069

**Comment by marcosnils** [comment]
@JacobLey  thx for taking this one. It seems to me that we need to do a bit more work here as IIUC schemas should be versioned with the engine. Apparently, we're only publishing a single version of the schema currently which I assume will cause issues given that we should set the same version as `engineVersion` has.

**Comment by JacobLey** [comment]
> So this issue is more around improving the DX right?

Right, my original issue was that my IDE had no clue what was "right", and the only way to test the correct format of the file is to run dagger. Adding the `$schema` field improved the feedback loop. Personally it is often easier to edit files locally than using CLI to configure them.

Then dagger was nuking that field when I tried running `dagger develop`, so I had to manually add it back. So I mostly wanted it to persist whatever I put there.

> given that we should set the same version as engineVersion

My _example_ is https://docs.dagger.io/reference/dagger.schema.json (which would always point to the latest, regardless of what version I am running locally), but my current proposed solution would accept https://not.a.real.website.com/not/a/real/schema.json just as well. Its more about the local IDE support rather than enforcing the provided JSON.

_If_ we were to bind the json schema to the engine version, or enforce that the json abides by the schema, would probably use a github link like https://raw.githubusercontent.com/dagger/dagger/refs/tags/v0.13.7/docs/current_docs/reference/dagger.schema.json which can point to that version's tag.

I would probably also expect the `$schema` to be automatically kept up to date as well, since that value is potentially out-of-date whenever you bump your engine version.
In the current proposed solution, it just allows the value to exist unchanged, but does not create or remove it.

**Comment by jedevc** [comment]
Yeah, okay, I think I agree here - if we have a versioned schema, then we really need to keep it up-to-date. But let's maybe continue that discussion in https://github.com/dagger/dagger/pull/9070, I don't think that blocks the main value of https://github.com/dagger/dagger/pull/9069.
